### ISSUE TITLE
What an Orphan is, exactly?

### ISSUE THREAD

**Issue by kopper** [openservicebrokerapi/servicebroker#617]
tl;dr; Is broker allowed to keep record of failed provisioning? If yes, then if Platform has a record of that Service Instance (e.g. with "provisioning failed" status) is this instance an Orphan or not? If not than some Platforms, esp. Service Catalog, don't have any Orphans (so why it implemented Orphan Mitigation in the first place?)

I'm not sure what (exactly) Orphan is. It is quite obvious to me, that if Platform doesn't have any record about Service Instance with instanceId=X, but Broker successfully provisioned X, then X is an Orphan.

Note that in Service Catalog that never happens (in normal operation), as it always requires ServiceInstance CRD to be created up front, even before it sends PUT request.

Here are some cases that are more interesting (using Service Catalog async provisioning as an example):

1. Broker has successfully provisioned an instance. Platform has a record of this Service Instance and...
a. doesn't know if provisioning has started (e.g. because of network issues didn't get any response from the Broker, or 5xx response from the Broker) => NOT an Orphan[1], Platform can ask later what the status is, and get valid response after network issues are resolved (current spec implies Orphan, because Orphan Mitigation is initiated)
b. thinks provisioning is in progress (= polling for PUT result in async provisioning in Service Catalog) => NOT an Orphan, will poll again soon
c. doesn't know if provisioning has completed, because polling was failing for long time due to network issues and eventually timed out => ??? it is unlikely that following DELETE request will succeed anyways, because something is wrong with the "network" (I don't even know what current spec says about this case)
d. thinks provisioning has failed (e.g. bug in broker which responded "failed" during Platform polling?) => ??? (can't see clear indication in the spec what it should be)

[1] - might become an Orphan after some time if polling was consistently failing

2. Broker has failed to provision an instance, but keeps Service Instance with some marker CREATE_FAILED or a similar record in some "instanceId table". Platform has a record of this Service Instance and...
a, b, c => the same as case 1
d. thinks provisioning has failed => ???

Is broker allowed to keep record of failed provisioning? It should keep it at least until Platform ACKs that it "learned" that state.

If it is allowed... is it an Orphan or not? If it is... how long? Can this state (Platform and Broker in sync pointing to a failed Service Instance) exists indefinitely or this instance should become an Orphan after some timeout?

I'd like to replace ??? with NOT an Orphan, because Platform has a record of that instance even if it is not in sync with the Broker. Maybe there should be a reconciliation process between Platform and Broker, but this is another story.

I'm happy to create a PR with clarifications, but first I'd like to get some feedback and validate my thoughts.

**Comment by duglin** [comment]
Couple of things to keep in mind...
- GET is optional for a broker to support - we need to make sure that our flows work even when GET isn't supported.
- we need to be backward compatible, which means any rules around mandating orphan mitigation can't be changed until we move to a v3.

> Is broker allowed to keep record of failed provisioning?

Sure, but IMO only within the confines of the spec. And by that I mean, it can keep a record of it, but after orphan mitigation is done a GET of that resource needs to return a 404 or 410 - even if the Broker has kept a "record" of it for logging/audit purposes.

> If yes, then if Platform has a record of that Service Instance (e.g. with "provisioning failed" status) is this instance an Orphan or not? If not than some Platforms, esp. Service Catalog, don't have any Orphans (so why it implemented Orphan Mitigation in the first place?)

I think the motivation here is that "we just don't know".  So, for example, just because the Platform see a 5xx error doesn't mean that the Broker sent it - it could have come from some middleware/proxy after the Broker successfully created the resource. So in this case we want the Platform to clean things up to make sure both sides agree that the resource is really gone.

**Comment by kopper** [comment]
> I think the motivation here is that "we just don't know"

If Platform has a record with some failure status of Service Instance, like Service Catalog does, is an instance like that considered Orphan? Let's put aside separate issue of recovering from failed provisioning and focus on preventing Orphans here. I'd argue that it depends on Platform implementation, e.g. Service Catalog has a record, so this instance is known to the platform, even if the status is out-of-sync (failed on Platform side, successfully provisioned in Broker). Platform can reconcile the status later, not necessarily be deleting the resource. Why delete it if it is healthy?

**Comment by mattmcneeney** [comment]
Hey @kopper 
Is this still an issue for you? Part of this problem was discussed further in #573 and I'm trying to understand if a problem still exists here.

**Comment by kopper** [comment]
Nope, this is not an issue for us anymore. We are not using Service Catalog
anymore.

**Comment by mattmcneeney** [comment]
Great, thank you!
